From 1ba920784dea851ef49150bf8cfaf56958952e93 Mon Sep 17 00:00:00 2001
From: "Guoniu.Zhou" <guoniu.zhou@nxp.com>
Date: Tue, 6 Feb 2018 11:34:13 +0800
Subject: [PATCH 3504/5242] MLK-17230-12: camera: fixed enum frame interval
 issue

commit  2961eb16f4b64c151d6800a21d4075e963f37f5a from
https://source.codeaurora.org/external/imx/linux-imx.git

when enum frame interval, kernel will be panic. It caused by
an null pointer, so correct the subdev data structure pointer.

ov5640 only support 15fps when mode is 1080P, so add this info
when user enum frame interval.

Reviewed-by: Sandor.Yu <sandor.yu@nxp.com>
Signed-off-by: Guoniu.Zhou <guoniu.zhou@nxp.com>
(cherry picked from commit 4089120ce5fbcb253728a23102a62c849b71c88b)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/imx8/mxc-mipi-csi2.c    |    2 +-
 drivers/media/platform/imx8/mxc-parallel-csi.c |    2 +-
 drivers/media/platform/imx8/ov5640_v3.c        |   25 ++++++++++++++----------
 3 files changed, 17 insertions(+), 12 deletions(-)

diff --git a/drivers/media/platform/imx8/mxc-mipi-csi2.c b/drivers/media/platform/imx8/mxc-mipi-csi2.c
index 992c3e8..8086819 100644
--- a/drivers/media/platform/imx8/mxc-mipi-csi2.c
+++ b/drivers/media/platform/imx8/mxc-mipi-csi2.c
@@ -577,7 +577,7 @@ static int mipi_csi2_enum_frame_interval(struct v4l2_subdev *sd,
 		return -EINVAL;
 	}
 
-	return v4l2_subdev_call(sd, pad, enum_frame_interval, NULL, fie);
+	return v4l2_subdev_call(sen_sd, pad, enum_frame_interval, NULL, fie);
 }
 
 static int mipi_csi2_get_fmt(struct v4l2_subdev *sd,
diff --git a/drivers/media/platform/imx8/mxc-parallel-csi.c b/drivers/media/platform/imx8/mxc-parallel-csi.c
index c6d5ddb..eeedffb 100644
--- a/drivers/media/platform/imx8/mxc-parallel-csi.c
+++ b/drivers/media/platform/imx8/mxc-parallel-csi.c
@@ -331,7 +331,7 @@ static int mxc_pcsi_enum_frame_interval(struct v4l2_subdev *sd,
 		return -EINVAL;
 	}
 
-	return v4l2_subdev_call(sd, pad, enum_frame_interval, NULL, fie);
+	return v4l2_subdev_call(sen_sd, pad, enum_frame_interval, NULL, fie);
 }
 
 static int mxc_pcsi_get_fmt(struct v4l2_subdev *sd,
diff --git a/drivers/media/platform/imx8/ov5640_v3.c b/drivers/media/platform/imx8/ov5640_v3.c
index 37b09ae..106cf8c 100644
--- a/drivers/media/platform/imx8/ov5640_v3.c
+++ b/drivers/media/platform/imx8/ov5640_v3.c
@@ -1146,7 +1146,7 @@ static int ov5640_enum_frameintervals(struct v4l2_subdev *sd,
 		struct v4l2_subdev_pad_config *cfg,
 		struct v4l2_subdev_frame_interval_enum *fie)
 {
-	int j, count;
+	int i, j, count;
 
 	if (fie->index < 0 || fie->index > ov5640_mode_MAX)
 		return -EINVAL;
@@ -1160,15 +1160,20 @@ static int ov5640_enum_frameintervals(struct v4l2_subdev *sd,
 	fie->interval.numerator = 1;
 
 	count = 0;
-	for (j = 0; j < (ov5640_mode_MAX + 1); j++) {
-		if (fie->width == ov5640_mode_info_data[j].width
-		 && fie->height == ov5640_mode_info_data[j].height
-		 && ov5640_mode_info_data[j].init_data_ptr != NULL) {
-			count++;
-		}
-		if (fie->index == (count - 1)) {
-			fie->interval.denominator = ov5640_framerates[0];
-			return 0;
+	for (i = 0; i < ARRAY_SIZE(ov5640_framerates); i++) {
+		for (j = 0; j < (ov5640_mode_MAX + 1); j++) {
+			if (fie->width == ov5640_mode_info_data[j].width
+			 && fie->height == ov5640_mode_info_data[j].height
+			 && ov5640_mode_info_data[j].init_data_ptr != NULL) {
+				count++;
+			}
+			if (fie->index == (count - 1)) {
+				fie->interval.denominator = ov5640_framerates[i];
+				if (ov5640_mode_info_data[j].mode ==
+						ov5640_mode_1080P_1920_1080)
+					fie->interval.denominator = ov5640_framerates[0];
+				return 0;
+			}
 		}
 	}
 
-- 
1.7.9.5

